soc/arm/atmel_sam/common/soc_poweroff.c

```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2023 Bjarki Arge Andreasen
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/sys/poweroff.h>
+#include <soc.h>
+
+#define SUPC_CR_KEY_PASSWORD (0xA5 << 24)
+
+void z_sys_poweroff(void)
+{
+	/* Switch off voltage regulator to enter backup mode */
+	SUPC->SUPC_CR = SUPC_CR_KEY_PASSWORD | SUPC_CR_VROFF_STOP_VREG;
```

Comment:
As stated at [hal_atmel/repository-specific-guidelines](https://github.com/zephyrproject-rtos/hal_atmel#repository-specific-guidelines) fixups should be send to hal.

Please, follow the guidelines at hal_atmel to send missing fixups and start to use
```diff
-	SUPC->SUPC_CR = SUPC_CR_KEY_PASSWORD | SUPC_CR_VROFF_STOP_VREG;
+	SUPC->SUPC_CR = SUPC_CR_KEY(SUPC_CR_KEY_PASSWD);
```

This is a partial implementation and not solve the other case that is when user design using the external power supply for VDDCODE. My initial reading is that a SUPC dts/driver should be added. I think at end you will need select the WKP sources too and maybe manage other topics like RTC.

I'm open to suggestion about this because it is not clear to me what kind of dts binding this probably will be. I don't expect a full driver and I believe in this case is start with the bare minimal and than incrementally add features.

Reply:
We could add it like
```
		supc: supc@400e1810 {
			compatible = "atmel,sam-supc";
			reg = <0x400e1810 0xd4>;
		};
```

and create some basic private APIs for entering backup mode and enabling gpio and component (RTC) wakeup events.

Then we can address it from the RTC driver if it is set as a `wakeup-source` (and to get the address of the SYSC_WPMR register)

```
		rtc: rtc@400e1810 {
			compatible = "atmel,sam-rtc";
			reg = <0x400e1810 0xd4>;
			supc = <&supc>;
		};
```

Reply:
What to do with GPIO wakeup events I'm not sure off yet. Maybe the gpio-controller could address the supc as well, and enable wakeup if we set some custom flag for the GPIO, like `ATMEL_SAM_WAKEUP` or something. but thats for later

Reply:
I think we can go the way you are proposing but let's keep GPIO WKP pins outside for now.
Do you agreed that there is space inside supc for sys_poweroff ?

Reply:
not exactly inside, sys_poweroff would call SUPC regulator VROFF (something like `supc_vreg_off()`), or the CMSIS DEEPSLEEP directly for the chips that require that :)